### PR TITLE
Convert to Triton Punica kernels

### PR DESCRIPTION
**Pull request by tgaddair**
- Add new Triton-based Punica kernels from vLLM that are faster and more flexible than existing Punica kernels
- Allow max_batch_prefill_tokens < max_input_length when chunked_prefill is enabled (chunk input sequence)
- Enable profiling when `LORAX_PROFILER_DIR` is set
- Fixed incompatibility between prefix caching and speculative decoding (https://github.com/huggingface/text-generation-inference/pull/2711)
- Fixed multi-Medusa to work with CUDA graphs
- Added Triton kernels to optimize batch preprocessing (https://github.com/huggingface/text-generation-inference/pull/2687)
- Trace preloaded adapter layers during initialization, instead of generic model defaults
- Remove flume (https://github.com/huggingface/text-generation-inference/pull/1184)
- Update Axum and router dependencies (https://github.com/huggingface/text-generation-inference/pull/1959)
- Fixed race condition causing stopped entries not to be removed from the cached batch

Breaking changes:
- Dynamic adapter loading and preloaded adapters can no longer be used together.

### CODE REVIEW

File: server/lorax_server/adapters/lora.py [hunk 1]
```diff
@@ -100,7 +100,8 @@ def __init__(
         self._is_transposed = False
 
         # [num_layers, hidden_size, r]
-        weights_a = [orient_for_rank(w, w.size(1)).contiguous() for w in weights_a]
+        # TODO(travis): add this back if rank is 8 and we're not using triton
+        # weights_a = [orient_for_rank(w, w.size(1)).contiguous() for w in weights_a]
```
Comment (tgaddair) [inline server/lorax_server/adapters/lora.py:104]: TODO: this needs to be fixed

File: server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py [hunk 2]
```diff
@@ -536,6 +537,10 @@ def forward(
 
         if lm_head_indices is not None:
             hidden_states = hidden_states[lm_head_indices]
+        
+        if skip_lm_head:
```
Comment (tgaddair) [inline server/lorax_server/models/custom_modeling/flash_qwen2_modeling.py:541]: TODO: this needs to be done for every model with an LM head

File: server/lorax_server/models/flash_causal_lm.py [hunk 3]
```diff
@@ -471,11 +547,15 @@ def filter(self, request_ids: List[int]) -> "FlashCausalLMBatch":
...
             adapter_meta = AdapterBatchMetadata(
                 adapter_indices=adapter_indices,
-                adapter_set=adapter_set,
+                adapter_list=adapter_list,
+                adapter_set=set(adapter_list),
                 adapter_segments=adapter_segments,
                 segment_indices=adapter_segment_indices,
             )
 
+        # logger.info("!!! FILTER slots {} -> {}", self.slots, slots)
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:556]: TODO: remove debug prints

File: server/lorax_server/models/flash_causal_lm.py [hunk 4]
```diff
@@ -679,18 +767,32 @@ def concatenate(cls, batches: List["FlashCausalLMBatch"]) -> "FlashCausalLMBatch
...
             adapter_segments, adapter_segment_indices = adapter_segment_builder.build()
             adapter_meta = AdapterBatchMetadata(
                 adapter_indices=adapter_indices,
+                adapter_list=adapter_list,
                 adapter_set=adapter_set,
                 adapter_segments=adapter_segments,
                 segment_indices=adapter_segment_indices,
             )
+        
+        # logger.info("!!! CONCATENATE slots {} -> {}", [b.slots for b in batches], slots)
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:794]: TODO

File: server/lorax_server/models/flash_causal_lm.py [hunk 5]
```diff
@@ -810,92 +957,116 @@ def prepare_for_prefill(self):
...
         self.adapter_meta = AdapterBatchMetadata(
             adapter_indices=adapter_indices,
-            adapter_set=adapter_set,
+            adapter_list=adapter_list,
+            adapter_set=set(adapter_list),
             adapter_segments=adapter_segments,
             segment_indices=adapter_segment_indices,
         )
 
+        # logger.info("!!! PREPARE_FOR_PREFILL slots {}", self.slots)
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:1067]: TODO

File: server/lorax_server/models/flash_causal_lm.py [hunk 6]
```diff
@@ -1340,6 +1525,11 @@ def forward(self, batch: FlashCausalLMBatch, adapter_data: AdapterBatchData) ->
         cache_lengths_tensor = batch.cache_lengths_tensor
         max_s = batch.max_current_length
 
+        # logger.info("!!! BLOCKS={} {}\n SLOTS={} {}\n SLOT_INDICES={} {}",
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:1528]: TODO

File: server/lorax_server/models/flash_causal_lm.py [hunk 7]
```diff
@@ -1422,6 +1627,10 @@ def forward(self, batch: FlashCausalLMBatch, adapter_data: AdapterBatchData) ->
                 lm_head_indices=batch.prefill_head_indices,
             )
 
+            if skip_lm_head:
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:1630]: TODO: check that model has an lm_head, otherwise assume we returned the logits

File: server/lorax_server/models/flash_causal_lm.py [hunk 8]
```diff
@@ -1919,13 +2140,12 @@ def generate_token(
                     batch.next_token_chooser.next_state(i, next_token_id)
 
             # Update values
+            # logger.info(f"!!! UPDATE VALUES {i} n_accepted_ids={n_accepted_ids} new_input_length={new_input_length} input_length={input_length} cache_length={cache_length}")
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:2143]: TODO

File: server/lorax_server/models/model.py [hunk 9]
```diff
@@ -246,6 +267,51 @@ def register_preloaded_adapters(
         )
         self.preloaded_adapters.extend(preloaded_adapters)
 
+        # For Triton kernels: need weights into contiguous tensor
+        # dict of (layer_name, layer_id) -> (lora_a_weights, lora_b_weights)
+        # where:
+        #   lora_a_weights = [num_adapters, r, hidden_size] 
+        #   lora_b_weights = [num_adapters, hidden_size, r]
+        for layer_name, layer_adapter_weights in self.layer_to_adapter_weights.items():
```
Comment (tgaddair) [inline server/lorax_server/models/model.py:275]: TODO: skip this is the triton punica kernels are disabled

File: server/lorax_server/server.py [hunk 10]
```diff
@@ -58,10 +58,14 @@ async def ServiceDiscovery(self, request, context):
...
-            self.cache.delete(request.id)
-        else:
-            self.cache.clear()
+        try:
+            if request.HasField("id"):
+                self.cache.delete(request.id)
+            else:
+                self.cache.clear()
+        except:
+            exit(1)
```
Comment (tgaddair) [inline server/lorax_server/server.py:67]: TODO: delete

File: server/lorax_server/utils/ops/utils.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import functools
```
Comment (tgaddair) [inline server/lorax_server/utils/ops/utils.py:1]: TODO: attribution for this module in the init file

File: router/src/server.rs [hunk 12]
```diff
@@ -1537,70 +1548,44 @@ pub async fn run(
...
-                .label("edge", edge);
-
-            let listener = tunnel.listen().await.unwrap();
-
-            // Run prom metrics and health locally too
-            tokio::spawn(
-                axum::Server::bind(&addr)
-                    .serve(
-                        Router::new()
-                            .route("/health", get(health))
```
Comment (tgaddair) [inline router/src/server.rs:1588]: @noyoshi @magdyksaleh not sure if this is needed and we need to add this back? Let me know.

**Review by ajtejankar — APPROVED** [review]

File: router/src/server.rs (hunk 12 above)
Comment (magdyksaleh) [reply router/src/server.rs:1588]: yes can we please keep it

Comment (tgaddair) [reply router/src/server.rs:1588]: Looks like this was only enabled when using ngrok tunneling, which is unsupported at the moment. Are we using ngrok today?
